docs/design/20190914.certificaterenewal.md

```diff
@@ -0,0 +1,170 @@
+- Implement watchers to wait for a specified period before marking these
+  certificates for renewal.
+- Implement a renewal process for rotating certificates for the target pod.
+- Reliably rediscover node certificates in the event of a driver restart on a
+  node.
+- Expose options for configuring the renewal process on a volume by volume basis
+  (disabling rotation, use the same private key, renew before duration).
+
+### Non-Goals
+
+- Require additional changes to upstream cert-manager to facilitate renewal.
+- Rely on the use of cert-manager's `Certificate` resource to facilitate renewal
+  functionality.
+- Implement any kind of Kubernetes controller behaviour through use of informers.
+
+## Proposal
+
+### Directory Structure
+
+In order to reliably discover and manage certificates for writing and watching,
+the directory structure needs to be consistent and deterministic for the driver.
+The main data directory will house multiple directories, each responsible for
+housing volume metadata as well as the mount point itself. It is possible that
+this main data directory contains other data created which should be successfully
+ignored by the driver.
+
+Each volume directory is named using the format of:
+
+```
+cert-manager-csi-$POD_NAMESPACE-$POD_NAME-$VOLUME_ID
+```
+
+This naming scheme ensures full namespacing of directory names whereby even if a
+single pod contains multiple cert-manager-csi volume mounts, each volume ID is
+ensured to be unique in this location. If anything external has created a
+directory name and causes a conflict, the driver should fail to create the
+volume until it has been resolved.
+
+Inside the pod directory contains the metadata file of the volume as well as the
+mount directory. Although the mount directory can have any child directory
+structure since a user can specify the key and certificate location, typically
+the directory will only contain 2 files at its root containing the key and
+certificate.
+
+```
+/cert-manager-csi-$POD_NAMESPACE-$POD_NAME-$VOLUME_ID/
+-- /metadata.json
+-- /data/
+---- /crt.pem
+---- /key.pem
+```
+
+### Meta Data File
+
+In order to retain information about the volume attributes, even if the driver
+has been restarted, data needs to be persisted on local storage. If this method
+was not taken, we would require to query the API server for all pods in the
+cluster, then sort the pods that have a cert-manager-csi volume attached and
+match by node. By keeping a local store, we eliminate our need for API server
+accesses to pods and reduce complexity. Volume mount attributes are immutable so
+the data shall not change during a pods life cycle.
+
+The contents of the meta data includes the volume attributes
+`map[string]string`, the volume ID and target path. With this a full recreation
+of the information for each volume can be made.
+
+The file will be stored as a `.json` file type. This is a widely used and
```

Comment:
Have we considered using the Kubernetes apimachinery for this? It will allow us to perform conversions in case we need to change things, and seems to be the way that most k8s components handle configuration etc on disk.

Reply:
Would be able to link something you're referring to?